### ISSUE TITLE
Add a property to [System.Management.Automation.ErrorRecord] to distinguish error types (non-terminating, statement-terminating, script-terminating)

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#4781]
As requested in #3158 and  suggested by @SteveL-MSFT 
 in https://github.com/PowerShell/PowerShell/issues/3647#issuecomment-327844929.

Currently, you cannot derive from a `[System.Management.Automation.ErrorRecord]` instance whether it represents a non-terminating, a statement-terminating, or a script-terminating error.

Note that, on the command line, you can already detect a _script_-terminating error generated with a  `Throw` statement by examining the relevant `$Error` collection entry's `.Exception.WasThrownFromThrowStatement` property, but that won't work for errors that were _promoted_ to script-terminating ones via `-ErrorAction Stop` / `$ErrorActionPreference = 'Stop'` (such errors generate an _accompanying_ `[System.Management.Automation.ActionPreferenceStopException]` error).

One useful application for allowing the discovery of a given error's "fatality" would be to enable error-type-specific coloring of error output, as discussed in the context of improving the standard error view in #3647.

Also see the expanded proposal [below](https://github.com/PowerShell/PowerShell/issues/4781#issuecomment-329585446), which asks for _two_ properties: the _intrinsic_ severity, and the _effective_ severity.

Environment data
----------------

```powershell
PowerShell Core v6.0.0-beta.6
```

**Comment by iSazonov** [comment]
With regard to coloring, we have a huge number of places where we write "error string" (not ErrorRecord) directly to the console. 
Only if we exclude the case we have a chance to implement the enhancement.

**Comment by mklement0** [comment]
Understood, but (a) colored console output is an important use case, and (b) in other cases we can potentially use _wording_ to indicate the error type.

**Comment by iSazonov** [comment]
Currently we do coloring in formatting system based on stream type only.
Perhaps it makes sense to make it more general and based on psobject type.

**Comment by mklement0** [comment]
Good point, and good idea.

**Comment by dragonwolf83** [comment]
As I mentioned in #3647, SQL Server differentiates terminating and non-terminating errors by calling it Severity. I think the term would be a good property name. Type will work too but Severity conveys a ranking of how an error will impact a script. 

It could make it easier to describe these error conditions in documentation when thinking in that term. For example, a non-terminating error is being acted as if it is less severe than a terminating error. Adding `ErrorAction -Stop` causes it to be treated with the same severity, a halt to the command.

**Comment by iSazonov** [comment]
Enough us to add `isTerminating` property?

**Comment by mklement0** [comment]
There are two distinct severities to consider:

* The _intrinsic_ severity of an error - what it is by design.

* The _effective_ severity - if it was _promoted to a script-terminating error_ or _suppressed_, by way of `-ErrorAction Stop` / `$ErrorActionPreference`.

While _severity_ is a good term to use, pragmatically speaking we _could_ get away with just an `IsTerminating` property, though perhaps `IsStatementTerminating` would be clearer (though I still wish this strange subdivision of terminating into script- and statement-terminating would never have happened - and perhaps was never intended to).

For _intrinsic_ severity:

* `.Exception.WasThrownFromThrowStatement` already identifies an intrinsically script-terminating error (generated with `Throw` - the only way to directly create a scrip-terminating error).

* `.IsStatementTerminating` containing `$True` would indicate a statement-terminating error, obviously.

* `.IsStatementTerminating` containing `$False` therefore indicates a _non-terminating_ error.

Looking at the above, perhaps `.Severity` with enum values isn't a bad idea.

---

For _effective_ severity:

* There is currently no way to tell if an error was _suppressed_, with `-ErrorAction SilentlyContinue` (`-ErrorAction Ignore` would prevent the error record from surfacing altogether). The same goes for a script/statement-terminating error caught with `Try` / `Catch`, though, arguably, [caught errors shouldn't surface either](https://github.com/PowerShell/PowerShell/issues/3768#issuecomment-323405734).

* Currently, the only way to tell if a non-terminating/statement-terminating error was _promoted_ to a script-terminating one is to examine the _very next_ `$Error` entry (`$Error[0]` immediately after the statement).

     * #4837 explains how this currently works, including how it works inconsistently and unhelpfully with statement-terminating errors.

---

Assuming it is technically feasible, it would be nice if both the intrinsic and the effective severity were reflected directly in the error record (though the separate `$Error` entry must probably be retained as well, for backward compatibility).

In summary, perhaps something along the lines of _two_ new properties:

`.Severity` (intrinsic), which can be one of `NonTerminating`, `StatementTerminating`, `ScriptTerminating`

`.EffectiveSeverity`, which can be one of the above, plus `Silenced`.

**Comment by BrucePay** [comment]
Something to keep in mind: there is, in fact, only one **kind** of ErrorRecord. It's what's done with that error record that varies. It can be written to the error pipeline, wrapped in a pipeline terminating exception and thrown or wrapped in an exception using the `throw` statement. The same error record can be used in multiple scenarios i.e. caught in try/catch and then written. The property should reflect this in some way.

**Comment by BrucePay** [comment]
Consider the case where an error goes from terminating to non-terminating:
```
 try { 1/0 } catch { write-error $_ }; "Hi"
```
In this example, 1/0 produces a terminating error that is caught and written to the error pipe as a non-terminating error. How would do you propose to flag that?

And how about:
```
PS[1] (109) > $e = try { 1/0 } catch { $_ }
PS[1] (110) > throw $e
```
In this case, the error is caught and simply written to stdout. It's not a terminating error anymore, it doesn't even show up in the error stream. Then in the next statement `throw $e` it's back to being a terminating error.

**Comment by mklement0** [comment]
As an aside: 

> It's not a terminating error anymore, it doesn't even show up in the error stream.

Currently, the error produced by ` $e = try { 1/0 } catch { $_ }` _does_ show up in the error stream, but it sounds like this will get fixed: #3768

**Comment by mklement0** [comment]
@BrucePay:

You're talking about the _effective_ severity - as ultimately surfaced, which is distinct from the _intrinsic_ severity - as originally emitted (see my [earlier comment](https://github.com/PowerShell/PowerShell/issues/4781#issuecomment-329585446)). Each piece of information of interest in different scenarios, so both should ideally be available as an `[ErrorRecord]` instance property.

As for a more concrete approach to tagging an error record with the _intrinsic_ severity:

* `Write-Error` / `$PSCmdlet.WriteError()` could tag the error as _non_-terminating.

* `$PSCmdlet.ThrowTerminatingError()` could tag the error as _statement_-terminating.

* Similarly, PowerShell itself all errors stemming from an expression (`1/0`) or .NET method call (`[int]::Parse('foo')`) or invalid command syntax (`Get-Item -NoSuchParam`) as _statement-terminating_

* The use of `Throw` could tag the error as _script_-terminating (fatal by default).

---

The _effective_ severity could be tagged as follows; if there are multiple layers, the top layer wins:

* An error record obtained from within a `catch` block should be tagged as _ignored_.
   * By contrast, one ignored with `-ErrorAction Ignore` would never even be seen.

* An error record processed in a `trap` block...
   *  ... that executes `break` should be tagged as _script_-terminating.
   *  ... that executes `continue` should be tagged as _silenced_.
   * Having neither `break` nor `continue` has no impact on the effective severity.

* An error record silenced via `-ErrorAction SilentlyContinue` / `$ErrorActionPreference = 'SilentlyContinue'` or `2>$null` should be tagged as _silenced_.

* An error record promoted via `-ErrorAction Stop` or `$ErrorActionPreference = 'Stop'` should be tagged as _script_-terminating.

* In all other cases, the effective severity would be the same as the intrinsic severity.
